main/src/tutorial/tutorial.flix

```diff
@@ -0,0 +1,237 @@
+// we consider is very simple, it has three types of statements:
+//
+//      x = c        [literal statement]
+//      x = y + z    [addition statement]
+//      x = y / z    [division statement]
+//
+//
+// We begin by defining a namespace for all our declarations. We could have
+// used the root namespace, but to avoid clashes with other examples we
+// create a new namespace.
+//
+namespace Tutorial.CP {
+
+    //
+    // We will represent the program-under-analysis using three input
+    // relations. One for each type of statement:
+    //
+    rel LitStm(r: Str, c: Int)
+    rel AddStm(r: Str, x: Str, y: Str)
+    rel DivStm(r: Str, x: Str, y: Str)
+
+    //
+    // Here the LitStm relation holds every literal statement of the form r = c.
+    // Similarly, for AddStm and DivStm.
+    //
+
+    //
+    // The next step is to define the lattice LocalVar which maps every local
+    // variable to an element of the constant propagation lattice (which we
+    // define below.)
+    //
+    // We can think of this lattice the result of the analysis.
+    //
+    lat LocalVar(k: Str, v: Constant)
+
+    //
+    // To define the constant propagation lattice, we begin by defining the
+    // type of constants. We define an enum with three variants: bottom and
+    // top elements, and a constructor Cst(i) for any integer i.
+    //
+    enum Constant {
+          case Top,
+
+        case Cst(Int),
+
+          case Bot
+    }
+
+    //
+    // The next task is to define the partial order on Constant.
+    // This is straightforward with pattern matching on the arguments:
+    //
+    def leq(e1: Constant, e2: Constant): Bool = match (e1, e2) with {
+        case (Constant.Bot, _)                      => true
+        case (Constant.Cst(n1), Constant.Cst(n2))   => n1 == n2
+        case (_, Constant.Top)                      => true
+        case _                                      => false
+    }
+
+    //
+    // Then its on to the least upper bound:
+    //
+    def lub(e1: Constant, e2: Constant): Constant = match (e1, e2) with {
+        case (Constant.Bot, x)                      => x
+        case (x, Constant.Bot)                      => x
+        case (Constant.Cst(n1), Constant.Cst(n2))   => if (n1 == n2) e1 else Constant.Top
+        case _                                      => Constant.Top
+    }
+
+    //
+    // ... and the greatest lower bound:
+    //
+    def glb(e1: Constant, e2: Constant): Constant = match (e1, e2) with {
+        case (Constant.Top, x)                      => x
+        case (x, Constant.Top)                      => x
+        case (Constant.Cst(n1), Constant.Cst(n2))   => if (n1 == n2) e1 else Constant.Bot
+        case _                                      => Constant.Bot
+    }
+
+    //
+    // This complete the specification of the lattice. To associate these
+    // functions with the type Constant we have to write some magic:
```

Comment:
Let's not use the terms "magic" or "weird syntax."

We could just say "... we have to use the following syntax (which is subject to change in the future)"